File: backend/src/routers/account.ts
import { Router, Response } from 'express';
import { requireAuth, AuthenticatedRequest } from '../auth/middleware';
import {
  getOrCreateSettings,
  updateSettings,
  listSavedCompanyBookmarks,
  addSavedCompany,
  removeSavedCompanyBookmark,
  getApiUsageSummary,
} from '../services/userService';

const router = Router();

// GET /api/v1/profile (user profile information)
router.get('/profile', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  return res.status(200).json({
    id: req.user!.id,
    clerk_user_id: req.user!.clerkUserId,
    email: req.user!.email,
    full_name: req.user!.fullName,
    avatar_url: req.user!.avatarUrl,
  });
});

// GET /api/v1/settings (user settings/preferences)
router.get('/settings', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const settings = await getOrCreateSettings(req.user!.id);
    return res.status(200).json({
      theme: settings.theme,
      preferred_llm_provider: settings.preferredLlmProvider,
      email_notifications: settings.emailNotifications,
    });
  } catch (error: any) {
    console.error('Get settings error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// PATCH /api/v1/settings (update preferences)
router.patch('/settings', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { theme, preferred_llm_provider, email_notifications } = req.body;
    const settings = await updateSettings(req.user!.id, {
      theme,
      preferredLlmProvider: preferred_llm_provider,
      emailNotifications: email_notifications,
    });

    return res.status(200).json({
      theme: settings.theme,
      preferred_llm_provider: settings.preferredLlmProvider,
      email_notifications: settings.emailNotifications,
    });
  } catch (error: any) {
    console.error('Update settings error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/v1/saved-companies (bookmarks list)
router.get('/saved-companies', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const companies = await listSavedCompanyBookmarks(req.user!.id);
    const formatted = companies.map(c => ({
      id: c.id,
      company_name: c.companyName,
      ticker: c.ticker,
      notes: c.notes,
      source: c.source,
      created_at: c.createdAt.toISOString(),
    }));
    return res.status(200).json(formatted);
  } catch (error: any) {
    console.error('List saved companies error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /api/v1/saved-companies (add bookmark)
router.post('/saved-companies', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { company_name, ticker, notes } = req.body;
    if (!company_name || typeof company_name !== 'string') {
      return res.status(400).json({ error: 'company_name is required and must be a string' });
    }

    const saved = await addSavedCompany(req.user!.id, company_name, ticker, notes);
    return res.status(200).json({
      id: saved.id,
      company_name: saved.companyName,
      ticker: saved.ticker,
      notes: saved.notes,
      created_at: saved.createdAt.toISOString(),
    });
  } catch (error: any) {
    console.error('Add saved company error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// DELETE /api/v1/saved-companies/:id (remove bookmark)
router.delete('/saved-companies/:id', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { id } = req.params;
    const success = await removeSavedCompanyBookmark(id, req.user!.id);

    if (!success) {
      return res.status(404).json({ error: 'Bookmark not found' });
    }

    return res.status(200).json({ message: 'Bookmark removed successfully' });
  } catch (error: any) {
    console.error('Delete saved company error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/v1/usage (quota/usage logs)
router.get('/usage', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const summary = await getApiUsageSummary(req.user!.id);
    return res.status(200).json(summary);
  } catch (error: any) {
    console.error('Get API usage error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend_python/app/routers/account.py
"""Profile, settings, saved companies, and API usage — powers the account pages."""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user
from app.database import get_db
from app.models import User
from app.schemas.user import (
    ApiUsageSummary,
    SavedCompanyCreate,
    SavedCompanyResponse,
    UserProfile,
    UserSettingsResponse,
    UserSettingsUpdate,
)
from app.services import (
    add_saved_company,
    get_api_usage_summary,
    get_or_create_settings,
    list_saved_companies,
    remove_saved_company,
    update_settings,
)

router = APIRouter(tags=["account"])


@router.get("/profile", response_model=UserProfile)
async def get_profile(user: User = Depends(get_current_user)) -> UserProfile:
    return UserProfile.model_validate(user)


@router.get("/settings", response_model=UserSettingsResponse)
async def get_settings_route(
    user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> UserSettingsResponse:
    settings_row = await get_or_create_settings(db, user)
    return UserSettingsResponse.model_validate(settings_row)


@router.patch("/settings", response_model=UserSettingsResponse)
async def update_settings_route(
    payload: UserSettingsUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> UserSettingsResponse:
    settings_row = await update_settings(
        db,
        user,
        theme=payload.theme,
        preferred_llm_provider=payload.preferred_llm_provider,
        email_notifications=payload.email_notifications,
    )
    return UserSettingsResponse.model_validate(settings_row)


@router.get("/saved-companies", response_model=list[SavedCompanyResponse])
async def get_saved_companies(
    user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> list[SavedCompanyResponse]:
    saved = await list_saved_companies(db, user.id)
    return [SavedCompanyResponse.model_validate(s) for s in saved]


@router.post("/saved-companies", response_model=SavedCompanyResponse, status_code=status.HTTP_201_CREATED)
async def create_saved_company(
    payload: SavedCompanyCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SavedCompanyResponse:
    saved = await add_saved_company(db, user.id, payload.company_name, payload.ticker, payload.notes)
    return SavedCompanyResponse.model_validate(saved)


@router.delete("/saved-companies/{saved_company_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_saved_company(
    saved_company_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    deleted = await remove_saved_company(db, saved_company_id, user.id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Saved company not found")


@router.get("/usage", response_model=list[ApiUsageSummary])
async def get_usage(
    user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> list[ApiUsageSummary]:
    summary = await get_api_usage_summary(db, user.id)
    return [ApiUsageSummary(**row) for row in summary]
